Add CharacterizeLinearTransform module

mikebind opened the pull request:
Finally following up on this suggestion: https://discourse.slicer.org/t/rotational-transformation-matrix-surface-registration-of-models/23680/5 .  Please let me know if any other changes or modifications are needed, thanks!

mikebind commented:
Also worked through my remaining issues for the StitchVolumes module referenced here: https://discourse.slicer.org/t/combine-two-volumes-into-one-using-python/23152/6

Since the first pull request (from just a few hours ago) hasn't been merged, this PR contains the changes for CharacterizeLinearTransform and for StitchVolumes.  If you want me to separate them somehow so they can be submitted separately please let me know.

lassoan commented on README.md:
```diff
@@ -58,3 +60,70 @@ If the slice size is increased then folding occurs:
+
+## Characterize Transform Matrix
+
+It is often difficult to understand what a transform matrix is doing just by inspection.  All the information is in those 12 numbers, but not in an easily understood format. CharacterizeTransformMatrix is a simple utility module which tries to quickly give you any information you might want to know about what a transformation matrix is doing.  For example, is it a rigid transformation or is there ...
```
```suggestion
It is often difficult to understand what a transform matrix is doing just by inspecting the numerical values.  All the information is in those 12 numbers, but not in an easily understood format. Characterize Transform Matrix module provides information on what a transformation matrix is doing.  For example, is it a rigid transformation or is there scaling?  If there is scaling, what are the scale factors and stretch directions?  Is there rotation?  If so, what is the axis of rotation and how much rotation occurs around that axis?  Alternatively, if we break down the rotation into a sequence of rotations around coordinate axes, what is the rotation about each axis?
```

lassoan commented on README.md:
```diff
@@ -58,3 +60,70 @@ If the slice size is increased then folding occurs:
+
+It is often difficult to understand what a transform matrix is doing just by inspection.  All the information is in those 12 numbers, but not in an easily understood format. CharacterizeTransformMatrix is a simple utility module which tries to quickly give you any information you might want to know about what a transformation matrix is doing.  For example, is it a rigid transformation or is there ...
+
+### To use
```
```suggestion
### How to use
```

lassoan commented on README.md:
````diff
@@ -58,3 +60,70 @@ If the slice size is increased then folding occurs:
+0 0 0 1 
+```
+### Some Decomposition Details
+This module uses polar decomposition to describe the components of a 4x4 transform matrix. The decomposition has the form: `H = T * R * K`, where `H` is the full homogeneous transformation matrix (with 0,0,0,1 as the bottom row), `T` is a translation-only matrix, `R` is a rotation-only matrix, and `K` is a stretch matrix. `K` can further be decompsed into three scale matrices, which can each be c ...
````
Can you justify the choice of polar decomposition? Why don't we use `vtkMath::Orthogonalize3x3`?

mikebind replied:
The clearest example I was following (I think from https://math.stackexchange.com/questions/237369/given-this-transformation-matrix-how-do-i-decompose-it-into-translation-rotati) used polar decomposition and that approach seemed appropriate.  It may be that an approach based on `vtkMath::Orthogonalize3x3` would also work, and I don't have a strong justification for why one approach might be better than the other.

lassoan commented on CharacterizeTransformMatrix/CharacterizeTransformMatrix.py:
```diff
@@ -0,0 +1,425 @@
+            R[:3, :3] = -R[:3, :3]
+            K[:3, :3] = -K[:3, :3]
+        # Check answer still OK
+        assert np.allclose(L, R @ K), "R*K should equal L, but it does not!"
```
`assert` is intended for testing and debugging. Please replace all `assert`s in non-test code by checking of the condition and raising an exception if the check fails.

mikebind replied:
Changed asserts to exceptions

lassoan commented on CharacterizeTransformMatrix/CharacterizeTransformMatrix.py:
```diff
@@ -0,0 +1,425 @@
+        #### Translation
+        translationVector = T[:3, 3]
+        line = (
+            "Lastly, this transformation translates, shifting:\n  %+0.1f mm in the R direction\n  %+0.1f mm in the A direction\n  %+0.1f mm in the S direction"
```
I would consider removing `Lastly, ` as it does not add any information.

mikebind replied:
Removed "Lastly" and added a final sentence about the order of the decomposed transformations

lassoan commented on CharacterizeTransformMatrix/CharacterizeTransformMatrix.py:
```diff
@@ -0,0 +1,425 @@
+            percentChange = 100 * (factor - 1)
+            percentChanges.append(percentChange)
+            scaleDirections.append(axis)
+            line = "  f%i: %+0.3f%% change in direction [%0.2f, %0.2f, %0.2f]" % (
```
Consider using markdown formatting. It makes the generated text more readable even if printed as text, and QTextEdit can render it nicely.

lassoan commented on CharacterizeTransformMatrix/CharacterizeTransformMatrix.py:
```diff
@@ -0,0 +1,425 @@
+        if np.abs(largestPercentChange) < 0.1:
+            isRigid = True
+            line = (
+                "This transform is essentially rigid (largest percent scale changes is %+0.3f%%, volume percent change is %+0.3f%%)."
```
I would recommend using f-strings to make the code more readable

mikebind replied:
Made the change to f-strings throughout

lassoan commented on StitchVolumes/StitchVolumes.py:
```diff
@@ -0,0 +1,706 @@
+    cropVolumeNode.SetIsotropicResampling(False)
+    cropVolumeNode.SetInterpolationMode(
+        cropVolumeNode.InterpolationNearestNeighbor
+    )  # use nearest neighbor to avoid resampling artifacts
```
Nearest neighbor introduces the worst artifacts. Linear or cubic needs to be used instead for better quality.

mikebind replied:
The motivation for this choice was to change voxel values as little as possible.  In the case where the new voxel grid is the same as the original voxel grids (same orientation and spacing), I felt that it was better to keep all of the original voxel values and translate them up to half a pixel onto the new grid than it would be to degrade the image by making every new voxel the average of 2 or 4 neighbors by linear interpolation.  If the new voxel grid differs from the original in orientation or spacing, then interpolation is clearly necessary and preferable to nearest neighbor.  In the earliest versions of this code, the orientation, spacing, and origin of the output grid was forced to be equal to that of the first input volume, and the grid was extended to minimally include all other input volumes.  In this case, I felt it was preferable to shift all other volumes in space slightly if their grid was slightly non-aligned rather than introduce interpolation. However, since the ROI is now decided upon independently from the input volumes, there is no guarantee of alignment with the first input volume grid (though this is still my use case for it) and mis-alignment definitely should be checked for and should trigger interpolation. I'll create an issue for this once PR is merged and and will try to work on it soon.

lassoan commented on StitchVolumes/StitchVolumes.py:
```diff
@@ -0,0 +1,706 @@
+        cropVolumeNode.InterpolationNearestNeighbor
+    )  # use nearest neighbor to avoid resampling artifacts
+    cropVolumeNode.SetFillValue(
+        0
```
I would recommend to compute fill value by vtkMRMLScalarVolumeNode::GetImageBackgroundScalarComponentAsDouble() (from the median of the 8 corner voxels of the volume)

mikebind replied:
I use the fill value being zero somewhere else in the logic, and I will need to fix that dependence before changing this.  I agree that it makes sense to base the value on the image corners rather than force to zero for all images.

lassoan commented on StitchVolumes/StitchVolumes.py:
```diff
@@ -0,0 +1,706 @@
+    ):
+        # Stitch together the supplied original volumes, resampling them
+        # into the space defined by the supplied roi, putting the stitched
+        # output into a volume with the given stitched volume name
```
Now that you add this nice and convenient module, it would be great if you could also make the algorithm a bit more sophisticated by computing weighted average of the volumes, using the distance from the volume edge (can be computed quickly using vtkITKDistanceTransform) as weighting factor.

mikebind replied:
It makes sense to add this as an option.  I would definitely retain the option to have a hard border to be able to easily evaluate alignment at the seams.  In addition, I would need to think about whether this could lead to averaging of image data with fill value data or not.

lassoan commented on StitchVolumes/StitchVolumes.py:
```diff
@@ -0,0 +1,706 @@
+            self._parameterNode.GetNodeReference("StitchedVolumeROI")
+            and self._parameterNode.GetNodeReference("InputVol1")
+            and self._parameterNode.GetNodeReference("InputVol2")
+            and self._parameterNode.GetParameter("OutputVolName")
```
For user convenience and conform with Slicer conventions, please create an output node selector. You can set the None text to "Create new volume" if you don't want to force the user create/select the output node manually.

mikebind replied:
Changed to node selector for output as suggested

lassoan reviewed (commented):
Thanks a lot, these are very useful modules!

Please add the two new modules to the extension (add the corresponding subdirectories in the top-level CMakeLists.txt file).

The changes contain a merge commit. Please rebase to remove it.

I've added a couple of more comments inline. They are not all mandatory for merging this pull request: you can submit an issue for each of them that you cannot address now.

mikebind commented:
> The changes contain a merge commit. Please rebase to remove it.

I'm not sure how to do this.  I see the git rebase command, and I think you are asking that I move all my commits to start from the current Sandbox state (instead of branching off from an earlier point), and also to just remove the merge commit (because it won't be relevant anymore)).  Do I do that on my fork (I think I can probably figure out how to do that) and then click "Contribute" again on my github fork to add those changes to this PR (I think that's what happens if I click that button again)?

mikebind commented:
OK, I tried the rebase and think I have successfully removed the merge commit. Is there anything else I need to do for that issue?  I will work through all the other suggested modifications now.

mikebind commented:
I have addressed all the comments raised in the review by @lassoan (thank you!)   There are a few remaining issues and feature requests which I will attempt to address over time.  I am not sure what the next step is, but I am going to try clicking to re-request a review because that seems logical :)

mikebind commented:
@lassoan, just bumping this.  Please let me know if there is anything which needs to be addressed before merging.

lassoan commented:
Sorry for the delay, I was travelling and many tasks piled up. I've merged your pull request now. Thanks a lot for your contributions!
